### DISCUSSION TITLE
Table issue

### DISCUSSION THREAD

**Discussion by bijlevel** [RomanKubiak/ctrlr#165] [Q&A]
I'm a bit reluctant to post this "issue" because I'm probably overlooking something VERY obvious... I have this code (snippet):
for i = 5,10 do							
			t1 = midi:getData():getByte(i);
			t3=BigInteger(t1)
			p=t3:getBitRangeAsInt(0,6) -- 6 bits, counting from the right to left. This effectively strips the first 2 bits
			Tbl = {[i-4] = p}
			console("p= "..Tbl[i-4])
		end
		for j =1,5 do
			console("table value=: "..Tbl[j])
		end
Within the "i "loop the console does show the values stored in the table as expected, but when it executes the "j" loop Ctrlr throws an error and tells me I'm concatenating a "nil" value, meaning the table is empty! How is that possible or what am I doing wrong?

**Comment by bijlevel** [comment]
Well, of course I went on trying and it worked after I changed the line "Tbl = {[i-4] = p}" to "Tbl[i-4] = p". Why it does work this way, I don't know.

**Reply by Ted-juh** [reply]
Tbl = { }   -- creates a empty table named Tbl

If you do Tbl = {[i-4]= p}, you are probably trying to fill the index i-4 with p. But here you are indexing [i-4] and not i-4. Not that it matters because you already figured out that you needed Tbl[i-4] = p.

for i = 5,10 do 
Tbl[i-4] = p 
end

You are telling to fill index 5-4/ 6-4/ 7-4/ 8-4/ 9-4 and 10-4 to be filled with the outcome of "p".
Is that what you want?

for j=1,5 do
console("table values = "..Tbl[j])
end

Well.. is there anything at values 1-5? I guess not and that's why it throws a nil error. 

for j=1,5 do
Tbl[j] = string.format("‎Parasaurolophus -%s",j)
console("table value"..string.format(" %s",j).." = "..Tbl[j])
end

Should give table index 1-5 a string  "‎Parasaurolophus-1" to "‎Parasaurolophus-5"  
But if I do -- console("value of Tbl[6]") -- it throws a nil error because that index isn't filled.

**Reply by bijlevel** [reply]
I think you're right. Indexing with [i-4] is indexing with a (kind of?) string, instead of what I expected with an index as the outcome of i-4 = 1. When using Tbl[i-4] the index starts (in my case) at 1, so with this way of indexing a table, the index is indeed 'calculated". Thanks for the explanation!

To learn/understand LUA better I'm using a text book. This book also writes that one could get also the value of a table by using the form **Tbl.j** where j is the index. But then Ctrlr throws an error again. So does this mean that Ctrlr is not 100% compatible with LUA?

**Comment by dnaldoog** [comment]
@bijlevel  If I was doing that, I would use _table.insert()_

<pre>
x=function(midi)
local Tbl = {} -- is Tbl global? - if so declare Tbl outside function
for i = 5, 10 do
    local t3 = BigInteger(midi:getData():getByte(i))
    local p = t3:getBitRangeAsInt(0, 6) -- 6 bits, counting from the right to left. This effectively strips the first 2 bits
    table.insert(Tbl, p)
    console("p= " .. Tbl[i - 4])
end
for j = 1, 5 do
    console("table value=: " .. Tbl[j])
end
end
</pre>

:)

**Reply by bijlevel** [reply]
Thank you, easier to implement and makes the code for sure more readable.

**Comment by dnaldoog** [comment]
The reason 

<pre>
for j =1,5 do
console("table value=: "..Tbl[j])
end
</pre>

doesn't work is because each time you run Tbl = {[i-4] = p} 
you are initialising a new table, so the only index that will end up existing in Tbl is Tbl[6] // i.e 10-4 = 6

The previous version of Tbl is destroyed on each iteration.

With regard to __Tbl.j__ this is accessing a hash based array, but by looping with an numeric iterator you can only access the index, not the key.

**Comment by dnaldoog** [comment]
Here is an explanation of tables. There are two types hash key based and indexed. They can be mixed
That is why you can do t={[0]="a","b","c","d"} -- mixed
for i=0,#t do
console(i.." "..t[i])
end

0 a
1 b
2 c
3 d

but ipairs will ignore the hash [0]

for i,v in ipairs (t) do
console(i.." "..v)
end

1 b
2 c
3 d

#t=3 _not_ 4
<pre>
--console=print -- uncomment for  https://repl.it/new/lua
---[[
Tbl={1="dnaldoog",2="bijlevel",3="tedjuh",4="goodweather",5="BAUS"} -- not legal

for i=1,5 do
console(Tbl[i])
end
  --]] 
--lua: [string "<eval>"]:1: '}' expected near '='


--do this instead, but it is not necessary

Tbl={[1]="dnaldoog",[2]="bijlevel",[3]="tedjuh",[4]="goodweather",[5]="BAUS"}

for i=1,5 do
console(Tbl[i])
end

-- same as 

Tbl={"dnaldoog","bijlevel","tedjuh","goodweather","BAUS"}

for i,v in ipairs(Tbl) do
console(v)
end
--[[
dnaldoog
bijlevel
tedjuh
goodweather
BAUS
--]]

--Hash array returns nil except BAUS which is indexed
Tbl={["1"]="dnaldoog",["2"]="bijlevel",["3"]="tedjuh",["4"]="goodweather",[5]="BAUS"}

for i=1,5 do
console(Tbl[i])
end
--[[
nil
nil
nil
nil
BAUS
--]]

Tbl={["1"]="dnaldoog",["s2"]="bijlevel",["3"]="tedjuh",["4"]="goodweather",["5"]="BAUS"}
or
Tbl={["1"]="dnaldoog",s2="bijlevel",["3"]="tedjuh",["4"]="goodweather",["5"]="BAUS"}
["1"]  distinguishes between 1

for k,v in pairs (Tbl) do
console(k.." "..v)
end

--[[
1   dnaldoog
3   tedjuh
s2   bijlevel
5   BAUS
4   goodweather
--]]
-- note in lua hash order is not the same as written returns 1 3 2 5 4

console (Tbl[1]) -- nil
console (Tbl.s2) -- bijlevel
-- console (Tbl.1) -- lua doesn't like this!!
</pre>

**Comment by Godlike-Productions** [comment]
Try to do this in a memoryBlock if you can

tempMB = MemoryBlock()
tempMB:createFromTable(tablename)
tempMB:append(next 6 bytes)

as example.

You can use all the getByte and getByteRange to pull things from MemoryBlocks.

I think MemoryBlocks are more efficient, the downside being that you'll need to convert them to something to display in the console (using something like toHexString())

I try to keep all binary (hex) data in memory blocks if I can.

**Reply by bijlevel** [reply]
memoryblocks being more efficient does it also automatically mean it is faster?

**Reply by Godlike-Productions** [reply]
I believe so, yes.  MemoryBlocks are essentially pointers to memory.  You don't need to parse in and out of tables

**Reply by dnaldoog** [reply]
I am always wondering about all this  this too _MemoryBlock()_ vs _t{}_, because _lua_ tables are supposed to be very fast; maybe it's all near enough not to be even noticeable unless the iterations are huge. 

The other one I was wondering about was which is faster __lua bit.band__ and friends or JUCE __BigInteger__? 

I believe an inbuilt function like _ipairs_ or _pairs_ is not as fast as 'for ' loops, but again the difference I am guessing is negligible.

**Reply by Ted-juh** [reply]
Follow-up question.. is table.insert/ table.remove for tables slower than just setByte for a memory block?

Reason is because I think I read somewhere on the old forums that strings and stuff were also supported with memoryBlock.

**Reply by Godlike-Productions** [reply]
table.insert and table.remove can be quite slow as the tables start getting bigger, and especially when they are complex structures.

I use MemoryBlock when I'm referring to a piece of data, such as a sysex string.  I put memory blocks into tables to store and manage them (ie each patch has the associated sysex data).

Tables are good for organization, MemoryBlock for data imho.

**Comment by dobo365** [comment]
Same for me. Most of the sysex manipulation is done with Memoryblocks.
However, I also use tables to define enums and this, jusyt in the classical way. 
For example:
`tCategories = {"[Not defined]", "Arpeggio", "Bass", "Brass", "Drone", "Keys", "Lead", "Noise", "Organ", "Other", "Pad", "Percussion", "Sound FX", "Sequence", "Strings", "Vox", "Woodwind"}`
